1.1

susiehgt opened the pull request:
This update contains: 

- UI improvements
   - Indication of mandatory fields (with "*") before each
   - When the user hover its mouse, the field required is indicated
   - Reorganisation of the display
   - Automatic updates
   
- On Experiment level:
   - Adding of variable experimental condition

- On interaction level: 
   - Adding of variable parameters

- On Participant Feature level:
   - Feature information selection from file
   - Adding of resulting sequence
   - Adding of feature role
   - Adding of feature xref qualifier
   - Adding of feature dependent variable parameters

- On the development side:
   - Removing of the multiple data enums to only keep one for easier maintenance
   - Refactoring of the utility classes for better separation of their role
   - Refactoring and renaming of the file processing classes 
   - Refactoring of the feature classes to separate the feature xrefs
   - Tests corrections
   - Javadoc improved
   - General code cleanup to avoid a maximum of warnings

Commit:
Adding of indication for mandatory fields;
Creation of parameters;
Adding of xref type;

Commit:
Adding of indication for mandatory fields;
Creation of parameters;
Adding of xref type;
adding of parameters on feature;
starting to refactoring to use objects instead of hashmap;
bug correction;

Commit:
Refactoring for cleaner and smaller code;
writing feature parameters to XML file;
working on adding resulting sequences on features;

Commit:
Adding resulting sequences on features;

Commit:
Selecting features from file;
Adding of tooltip for user;

Commit:
Selecting features from file;
Adding of tooltip for user;

Commit:
Cleanup

Commit:
Correction of the display;
Adding previous id and db in the selection for uniprot mapping

Commit:
Adding of properties file

Commit:
Adding of properties file

Commit:
Automatic update: OK on mac, to test on windows and linux;
Prochains handled correctly;
Correction of file formatting bug;
Fetching feature xrefs from file;
Correction of feature creation between baits and preys;

Commit:
Adding of feature role;
Cleanup of the code;
Debug for csv, tsv files;
Adding of documentation;

Commit:
Correcting tests following the latests changes;
Code cleanup;
Javadoc;

Commit:
Cleaning up feature creation code;
Fixing feature type and xref type bug;

Commit:
Code cleanup;
TODO: check for the feature xref input when selected from file which does not work

Commit:
Code cleanup;
Feature selected from file and user input bug fix;
Reorganising utils classes;

Commit:
Merging DataForRawFile.java and DataTypeAndColumn.java into InputData.java;
Refactoring:
- File processing classes
- Utility classes
- Feature classes

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/gui/FileFormaterGui.java:
```diff
@@ -104,26 +119,30 @@ private void setupComboBoxDefaults() {
         baitColumn.removeAllItems();
-        baitColumn.addItem("Select baits column");
+        baitColumn.setToolTipText("* Select baits id column");
+        baitColumn.addItem("* Select baits id column");
         baitColumn.setEnabled(true);
 
         preyColumn.removeAllItems();
-        preyColumn.addItem("Select preys column");
+        preyColumn.setToolTipText("* Select preys id column");
+        preyColumn.addItem("* Select preys id column");
```
Why do these items have a * at the beginning, but the other ones below don't?

susiehgt replied:
It's to indicate that they are mandatory

jmedinaebi commented:
This is a long PR, can you add in the description what are the changes in this version?

Commit:
Debug uniprot mapper;

Commit:
Debug uniprot mapper;
Code cleanup;

Commit:
Debug feature creation

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/FileWriter.java:
```diff
@@ -110,27 +110,33 @@ public void writeToExcel(List<Map<String, String>> data, String filePath, String
         if (idColumnIndex == -1) {
-            showErrorDialog("ID column not found: " + idColumnName);
-            LOGGER.severe("ID column not found: " + idColumnName);
+            showErrorDialog("ID column not found: " + idColumnIndex);
```
Does it make sense to show the column index instead of column name when displaying an error?

susiehgt replied:
I'm not working with the column name anymore, so if it is not found we don't have how it's called, just what it was supposed to be... I removed the index in the message because it was useless

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/FileWriter.java:
```diff
@@ -274,6 +289,8 @@ private void processRow(List<String> row, int idColumnIndex, int previousIdDbCol
      * @param lastCellIndex      Index at which to start writing new UniProt-related cells.
      */
     private void processWorkbookRow(Row row, int idColumnIndex, int idDbColumnIndex, int organismColumnIndex, int lastCellIndex) {
+        System.out.println(idColumnIndex + " " + idDbColumnIndex + " " + organismColumnIndex);
```
Remember to remove these type of debugging before merging and releasing this.

Commit:
Following comments on PR#1:
- correction of error messages for id column not found
- removing unnecessary print

jmedinaebi commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
   <groupId>uk.ac.ebi.intact.psi.mi.xmlmaker</groupId>
   <artifactId>PSI-MI-XML-maker</artifactId>
-  <version>1.0</version>
+  <version>1.1-SNAPSHOT</version>
```
It should be 1.1 when we're ready to merge it.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/XmlMakerGui.java:
```diff
@@ -30,13 +35,13 @@
 
-    private static final int FRAME_WIDTH = 2000;
-    private static final int FRAME_HEIGHT = 2000;
+    private static final int FRAME_WIDTH = getDefaultToolkit().getScreenSize().width - 50;
```
What's 50? If it's some margin we wanna leave for some reason, we should do:
```
private static final int SOME_MARGIN = 50; // Maybe not named SOME_MARGIN but something relevant that explains it
private static final int FRAME_WIDTH = getDefaultToolkit().getScreenSize().width - SOME_MARGIN;
private static final int FRAME_HEIGHT = getDefaultToolkit().getScreenSize().height - SOME_MARGIN;
```

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/XmlMakerGui.java:
```diff
@@ -426,6 +439,8 @@ private JButton createRestartButton(JFrame currentFrame) {
     private JPanel createFileFormaterPanel(){
         JPanel fileFormaterPanel = fileFormaterGui.getFileFormaterPanel();
         fileFormaterPanel.setAutoscrolls(true);
+        fileFormaterPanel.setSize(new Dimension(FRAME_WIDTH, 350));
```
We could also move the 350 to a static variable with a relevant name.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/FileFormater.java:
```diff
@@ -86,235 +107,202 @@ public void selectFileFormater(String filePath,
 
+        try {
+            iterator.remove();
+        } catch (UnsupportedOperationException ignored) {
```
Why are we ignoring this exception?

susiehgt replied:
I think I had a warning at some points but I removed it and it seems to not be here anymore

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/FileFormater.java:
```diff
@@ -86,235 +107,202 @@ public void selectFileFormater(String filePath,
+                    oneParticipant.put(field.name, interactionData.get(field.name + BAIT.name));
+                } else if ("prey".equalsIgnoreCase(experimentalRole)) {
+                    oneParticipant.put(field.name, interactionData.get(field.name + PREY.name));
+                }
```
These ifs and elses look confusing, we could change it to:
```
            if (field.initial) {
                if (field.experimentalRoleDependent) {
                    if ("bait".equalsIgnoreCase(experimentalRole)) {
                        oneParticipant.put(field.name, interactionData.get(field.name + BAIT.name));
                    } else if ("prey".equalsIgnoreCase(experimentalRole)) {
                        oneParticipant.put(field.name, interactionData.get(field.name + PREY.name));
                    }
                } else {
                    oneParticipant.put(field.name, interactionData.get(field.name));
                }
            }
```

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/FileFormater.java:
```diff
@@ -359,37 +337,92 @@ private void addAllFeatures() {
+     */
+    private int numberOfFeatureCells(){
+        int total = 0;
+        for (InputData data : InputData.values()) {
```
```
return Arrays.stream(InputData.values()).filter(data -> !data.initial).count();
```

I think that would be the same, although it may be a long so you need to cast it to int.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/content/InputData.java:
```diff
@@ -0,0 +1,108 @@
+    @Getter @Setter
+    private int index;
+
+    InputData(String name, boolean experimentalRoleDependent, boolean initial) {
```
I think if we use `@RequiredArgsConstructor` Lombok would generate this exact constructor.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/content/InputData.java:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+        return initialData;
+    }
```
We could use streams with filters and collect to List for these methods.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/file/processing/gui/FileFormaterGui.java:
```diff
@@ -21,25 +26,28 @@
     private final JComboBox<String> preyNameColumn = new JComboBox<>();
 
-    final ParticipantAndInteractionCreatorGui participantAndInteractionCreatorGui = new ParticipantAndInteractionCreatorGui();
+    private final int WIDTH = getDefaultToolkit().getScreenSize().width - 100;
```
We could also define the height here, we're currently using height as 100 in multiple places in this class, and if we want to change we need to change it in all the places. Let's have it here just defined once.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/jami/creators/FeatureData.java:
```diff
@@ -0,0 +1,22 @@
+package uk.ac.ebi.intact.psi.mi.xmlmaker.jami.creators;
+
+public class FeatureData {
```
The members should be private and use getters and setters. Not particularly relevant, so don't worry much if takes too much time to change it.

jmedinaebi replied:
We could also use `@Builder` if you prefer.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/jami/creators/XmlFeatureEvidenceCreator.java:
```diff
@@ -0,0 +1,332 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+//import static uk.ac.ebi.intact.psi.mi.xmlmaker.file.processing.content.DataTypeAndColumn.*;
```
Delete.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/jami/creators/XmlFeatureEvidenceCreator.java:
```diff
@@ -0,0 +1,332 @@
+            return PositionUtils.createUndeterminedPosition();
+        }
+
+        range = range.trim().replaceAll(";$", "");
```
Why do we need this replace? If there's a known reason why there may be a `;` at the end of the ranges, we could add a comment about it.

susiehgt replied:
Yes we need to replace it for the xml writing because it stays and can cause issues when we are fetching for the mi- terms, I will add a comment about it

susiehgt replied:
I removed it in the end and modified where those cases were treated.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/jami/gui/InteractionsCreatorGui.java:
```diff
@@ -222,7 +225,7 @@ protected void setValue(Object value) {
 
     /**
-     * Configures the overall appearance of the table, including header and resizing behaviour.
+     * Configures the overall appearance of the table, including header and resizing behavior.
```
Did IntelliJ complain about this? `behaviour` is the British spelling and `behavior` is the American spelling. I think IntelliJ always complains about spelling mistakes like this.

susiehgt replied:
Yes I tried to remove as much as I could from the code analysis

jmedinaebi reviewed (commented):
I haven't gone through the whole review yet. It's looking good, most of my comments are about using Lombok more, static variables and streams.

One thing I haven't written in any other comment. You have some private final variables in different classes, any reason for those not to be static? You also have some final variables without the private, any reason for that?

Commit:
Following comments on PR#1:
- code cleanup and simplifying;

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/utils/XmlMakerUtils.java:
```diff
@@ -45,23 +46,35 @@ public class XmlMakerUtils {
         }
     }
 
+    //Connection
```
Remove this comment, as we already have a javadoc comment for this method.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/utils/VersionUtils.java:
```diff
@@ -0,0 +1,225 @@
+public class VersionUtils {
+
+    private static final Logger LOGGER = Logger.getLogger(VersionUtils.class.getName());
+//    private static final String VERSION_URL = "https://raw.githubusercontent.com/MICommunity/psi-mi-xml-maker/refs/heads/main/src/main/resources/xmlMaker.properties"; //todo: change for release
+    private static final String VERSION_URL = "https://raw.githubusercontent.com/MICommunity/psi-mi-xml-maker/refs/heads/1.1/src/main/resources/xmlMaker.properties";
```
Remember to update the version URL to the main branch when we're ready to merge.

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/utils/VersionUtils.java:
```diff
@@ -0,0 +1,225 @@
+    private static final Logger LOGGER = Logger.getLogger(VersionUtils.class.getName());
+//    private static final String VERSION_URL = "https://raw.githubusercontent.com/MICommunity/psi-mi-xml-maker/refs/heads/main/src/main/resources/xmlMaker.properties"; //todo: change for release
+    private static final String VERSION_URL = "https://raw.githubusercontent.com/MICommunity/psi-mi-xml-maker/refs/heads/1.1/src/main/resources/xmlMaker.properties";
+    private static final String GITHUB_URL = "https://github.com/MICommunity/psi-mi-xml-maker/releases/download/";
```
This link gets me to a 404. Is it the right URL?

jmedinaebi replied:
Nevermind, I just saw how we use it, ignore my previous comment.

jmedinaebi commented:
Just comments about the version number and github URL, to make the changes before merging. Once Kalpana has tested the latest changes, let's change it and then we can merge it, and create the new release in Github.

Commit:
Bug fix for variable experimental prep;

jmedinaebi reviewed (commented):
Is everything now working after the last fix? If so, please update the version and URL and then we can merge this.

susiehgt commented:
> Is everything now working after the last fix? If so, please update the version and URL and then we can merge this.

Kalpana wanted to test it one last time, but it should be okay

Commit:
Variable experimental params referencing on the interactions

Commit:
Change of version URL link

Commit:
Change of version in pom

jmedinaebi commented on src/main/java/uk/ac/ebi/intact/psi/mi/xmlmaker/jami/XmlFileWriter.java:
```diff
@@ -227,6 +227,7 @@ private void writeInteractionsToFile(List<XmlInteractionEvidence> interactions,
             showErrorDialog("Error during PSI-XML writing, please check that the columns are correctly associated" +
                     "and that the file is correctly formatted. \n" + e.getMessage());
             LOGGER.error("Error during PSI-XML writing", e);
+            e.printStackTrace();
```
Do we want to keep this?
